server/src/main/java/org/opensearch/cluster/store/RemoteClusterStateService.java

```diff
@@ -0,0 +1,156 @@
+package org.opensearch.cluster.store;
+
+import static org.opensearch.indices.IndicesService.CLUSTER_REMOTE_STATE_REPOSITORY_SETTING;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Supplier;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.cluster.ClusterState;
+import org.opensearch.cluster.metadata.IndexMetadata;
+import org.opensearch.cluster.store.ClusterMetadataMarker.UploadedIndexMetadata;
+import org.opensearch.common.blobstore.BlobContainer;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.indices.IndicesService;
+import org.opensearch.repositories.RepositoriesService;
+import org.opensearch.repositories.Repository;
+import org.opensearch.repositories.blobstore.BlobStoreRepository;
+import org.opensearch.repositories.blobstore.ChecksumBlobStoreFormat;
+
+/**
+ * A Service which provides APIs to upload and download cluster metadata from remote store.
+ *
+ * @opensearch.internal
+ */
+public class RemoteClusterStateService {
```

Comment:
Should this be an alternate implementation of PersistedClusterStateService#Writer, we can call this a `RemoteWriter`?

Reply:
PersistedClusterStateService#Writer is not an interface. It is a concrete class which has logic for writing to Lucene index. I think we should a separate class rather than extending Writer class

Reply:
I understand that the current state might not be extensible. The suggestion was to refactor and use a common `Writer` interface for both local and remote state for consistency reasons.

Reply:
A new common interface for PersistedClusterStateService.Writer and RemoteClusterStateService write methods should be possible. But it would need a lot of refactor in PersistedClusterStateService and its unit tests. Let me create an issue to track this separately.